File: c/hangman.c
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define MAX_WORD_LENGTH 50
#define MAX_TRIES 6

// Function prototypes
void displayWelcome();
void displayMenu();
void displayHangman(int wrongGuesses);
void displayWord(const char* word, const char* guessed, int length);
int playGame(const char* word);
void displayGameOver(int won, const char* word);
char getGuess(const char* alreadyGuessed);
int isLetterInWord(char letter, const char* word);
void addToGuessed(char* guessed, char letter);
int hasWon(const char* word, const char* guessed);

// Word bank for the game
const char* wordBank[] = {"programming", "computer",  "keyboard",  "function",
                          "variable",    "algorithm", "database",  "network",
                          "software",    "hardware",  "developer", "debugging",
                          "compiler",    "memory",    "processor"};

int main() {
  int choice;
  int playAgain = 1;

  srand(time(NULL));  // Seed random number generator

  displayWelcome();

  while (playAgain) {
    displayMenu();
    printf("Enter your choice: ");
    scanf("%d", &choice);
    getchar();  // Clear newline

    switch (choice) {
      case 1: {
        // Play with random word
        int randomIndex = rand() % (sizeof(wordBank) / sizeof(wordBank[0]));
        const char* selectedWord = wordBank[randomIndex];
        int won = playGame(selectedWord);
        displayGameOver(won, selectedWord);
        break;
      }
      case 2: {
        // Play with custom word
        char customWord[MAX_WORD_LENGTH];
        printf("\nEnter a word for player 2 to guess: ");
        fgets(customWord, MAX_WORD_LENGTH, stdin);
        customWord[strcspn(customWord, "\n")] = 0;  // Remove newline

        // Convert to lowercase
        for (int i = 0; customWord[i]; i++) {
          customWord[i] = tolower(customWord[i]);
        }

// Clear screen (works on most systems)
#ifdef _WIN32
        system("cls");
#else
        system("clear");
#endif

        int won = playGame(customWord);
        displayGameOver(won, customWord);
        break;
      }
      case 3:
        printf("\nThanks for playing! Goodbye!\n");
        playAgain = 0;
        break;
      default:
        printf("\nInvalid choice! Please try again.\n\n");
    }

    if (playAgain && choice != 3) {
      printf("\nPlay again? (1 = Yes, 0 = No): ");
      scanf("%d", &playAgain);
      getchar();  // Clear newline
      printf("\n");
    }
  }

  return 0;
}

void displayWelcome() {
  printf("\n");
  printf("========================================\n");
  printf("    WELCOME TO HANGMAN GAME!\n");
  printf("========================================\n");
  printf("\nGuess the word one letter at a time.\n");
  printf("You have %d wrong guesses before you lose!\n\n", MAX_TRIES);
}

void displayMenu() {
  printf("========================================\n");
  printf("             MAIN MENU\n");
  printf("========================================\n");
  printf("1. Play with Random Word\n");
  printf("2. Play with Custom Word (2 Players)\n");
  printf("3. Exit\n");
  printf("========================================\n");
}

void displayHangman(int wrongGuesses) {
  printf("\n");
  printf("  +---+\n");
  printf("  |   |\n");

  if (wrongGuesses >= 1) {
    printf("  O   |\n");
  } else {
    printf("      |\n");
  }

  if (wrongGuesses >= 4) {
    printf(" /|\\  |\n");
  } else if (wrongGuesses >= 3) {
    printf(" /|   |\n");
  } else if (wrongGuesses >= 2) {
    printf("  |   |\n");
  } else {
    printf("      |\n");
  }

  if (wrongGuesses >= 6) {
    printf(" / \\  |\n");
  } else if (wrongGuesses >= 5) {
    printf(" /    |\n");
  } else {
    printf("      |\n");
  }

  printf("      |\n");
  printf("=========\n");
  printf("\nWrong guesses: %d / %d\n", wrongGuesses, MAX_TRIES);
}

void displayWord(const char* word, const char* guessed, int length) {
  printf("\nWord: ");
  for (int i = 0; i < length; i++) {
    if (strchr(guessed, word[i]) != NULL) {
      printf("%c ", word[i]);
    } else {
      printf("_ ");
    }
  }
  printf("\n");
}

int playGame(const char* word) {
  int wordLength = strlen(word);
  char guessed[27] = "";  // Store guessed letters
  int wrongGuesses = 0;
  int won = 0;

  printf("\n========================================\n");
  printf("         GAME STARTED!\n");
  printf("========================================\n");

  while (wrongGuesses < MAX_TRIES && !won) {
    displayHangman(wrongGuesses);
    displayWord(word, guessed, wordLength);

    if (strlen(guessed) > 0) {
      printf("\nGuessed letters: %s\n", guessed);
    }

    char guess = getGuess(guessed);

    if (isLetterInWord(guess, word)) {
      printf("\nCorrect! '%c' is in the word.\n", guess);
      addToGuessed(guessed, guess);

      if (hasWon(word, guessed)) {
        won = 1;
      }
    } else {
      printf("\nWrong! '%c' is not in the word.\n", guess);
      addToGuessed(guessed, guess);
      wrongGuesses++;
    }

    printf("\n");
  }

  // Display final state
  displayHangman(wrongGuesses);
  displayWord(word, guessed, wordLength);

  return won;
}

char getGuess(const char* alreadyGuessed) {
  char guess;
  int valid = 0;

  while (!valid) {
    printf("\nEnter your guess (a-z): ");
    scanf(" %c", &guess);
    getchar();  // Clear newline

    guess = tolower(guess);

    if (!isalpha(guess)) {
      printf("Invalid input! Please enter a letter.\n");
    } else if (strchr(alreadyGuessed, guess) != NULL) {
      printf("You already guessed '%c'! Try another letter.\n", guess);
    } else {
      valid = 1;
    }
  }

  return guess;
}

int isLetterInWord(char letter, const char* word) {
  return strchr(word, letter) != NULL;
}

void addToGuessed(char* guessed, char letter) {
  int len = strlen(guessed);
  guessed[len] = letter;
  guessed[len + 1] = '\0';
}

int hasWon(const char* word, const char* guessed) {
  for (int i = 0; word[i] != '\0'; i++) {
    if (strchr(guessed, word[i]) == NULL) {
      return 0;
    }
  }
  return 1;
}

void displayGameOver(int won, const char* word) {
  printf("\n========================================\n");
  if (won) {
    printf("    CONGRATULATIONS! YOU WON!\n");
    printf("========================================\n");
    printf("\nYou guessed the word: %s\n", word);
  } else {
    printf("        GAME OVER!\n");
    printf("========================================\n");
    printf("\nThe word was: %s\n", word);
    printf("Better luck next time!\n");
  }
}

File: python/hangman.py
"""
Hangman Game - Main Module
A text-based interactive Hangman game with OOP design.
"""

import random
import os


class HangmanDisplay:
    """Handles the visual representation of the hangman."""

    STAGES = [
        """
           --------
           |      |
           |
           |
           |
           |
           -
        """,
        """
           --------
           |      |
           |      O
           |
           |
           |
           -
        """,
        """
           --------
           |      |
           |      O
           |      |
           |
           |
           -
        """,
        """
           --------
           |      |
           |      O
           |     \\|
           |
           |
           -
        """,
        """
           --------
           |      |
           |      O
           |     \\|/
           |
           |
           -
        """,
        """
           --------
           |      |
           |      O
           |     \\|/
           |      |
           |
           -
        """,
        """
           --------
           |      |
           |      O
           |     \\|/
           |      |
           |     / \\
           -
        GAME OVER!
        """,
    ]

    @staticmethod
    def get_stage(wrong_guesses):
        """Return the hangman stage based on wrong guesses."""
        stage_index = min(wrong_guesses, len(HangmanDisplay.STAGES) - 1)
        return HangmanDisplay.STAGES[stage_index]


class WordBank:
    """Manages the word list and word selection."""

    WORD_LIST = [
        "python",
        "javascript",
        "programming",
        "developer",
        "algorithm",
        "function",
        "variable",
        "database",
        "interface",
        "framework",
        "computer",
        "software",
        "hardware",
        "network",
        "security",
        "encryption",
        "debugging",
        "repository",
        "deployment",
        "terminal",
        "abstract",
        "inheritance",
        "polymorphism",
        "encapsulation",
        "recursion",
        "iteration",
        "compilation",
        "exception",
        "middleware",
    ]

    @staticmethod
    def get_random_word():
        """Return a random word from the word bank."""
        return random.choice(WordBank.WORD_LIST).upper()

    @staticmethod
    def add_words(words):
        """Add new words to the word bank."""
        WordBank.WORD_LIST.extend(words)


class Word:
    """Represents the secret word to be guessed."""

    def __init__(self, word):
        self.word = word.upper()
        self.letters = set(self.word)
        self.guessed_letters = set()

    def guess_letter(self, letter):
        """
        Add a guessed letter and return True if correct, False otherwise.
        """
        letter = letter.upper()
        self.guessed_letters.add(letter)
        return letter in self.letters

    def is_solved(self):
        """Check if the word has been completely guessed."""
        return self.letters.issubset(self.guessed_letters)

    def get_display(self):
        """Return the word with unguessed letters as underscores."""
        return " ".join(
            letter if letter in self.guessed_letters else "_" for letter in self.word
        )

    def get_word(self):
        """Return the original word."""
        return self.word


class Player:
    """Represents a player in the game."""

    def __init__(self, name="Player"):
        self.name = name
        self.score = 0
        self.games_played = 0
        self.games_won = 0

    def add_win(self):
        """Record a win for the player."""
        self.games_won += 1
        self.games_played += 1

    def add_loss(self):
        """Record a loss for the player."""
        self.games_played += 1

    def add_score(self, points):
        """Add points to player's score."""
        self.score += points

    def get_stats(self):
        """Return player statistics."""
        win_rate = (
            (self.games_won / self.games_played * 100) if self.games_played > 0 else 0
        )
        return {
            "name": self.name,
            "score": self.score,
            "games_played": self.games_played,
            "games_won": self.games_won,
            "win_rate": win_rate,
        }


class HangmanGame:
    """Main game controller for Hangman."""

    MAX_WRONG_GUESSES = 6

    def __init__(self, player=None):
        self.player = player if player else Player()
        self.word = None
        self.wrong_guesses = 0
        self.game_over = False
        self.won = False

    def start_new_game(self):
        """Initialize a new game."""
        self.word = Word(WordBank.get_random_word())
        self.wrong_guesses = 0
        self.game_over = False
        self.won = False

    def make_guess(self, letter):
        """
        Process a letter guess.
        Returns a tuple: (is_correct, message)
        """
        if not letter or len(letter) != 1 or not letter.isalpha():
            return None, "Please enter a single letter."

        letter = letter.upper()

        if letter in self.word.guessed_letters:
            return None, f"You already guessed '{letter}'. Try a different letter."

        is_correct = self.word.guess_letter(letter)

        if is_correct:
            if self.word.is_solved():
                self.won = True
                self.game_over = True
                self.player.add_win()
                points = (self.MAX_WRONG_GUESSES - self.wrong_guesses) * 10
                self.player.add_score(points)
                return True, f"Correct! You won! +{points} points"
            return True, f"Good guess! '{letter}' is in the word."
        else:
            self.wrong_guesses += 1
            if self.wrong_guesses >= self.MAX_WRONG_GUESSES:
                self.game_over = True
                self.player.add_loss()
                return False, f"Wrong! The word was: {self.word.get_word()}"
            return False, f"Sorry, '{letter}' is not in the word."

    def get_game_state(self):
        """Return the current game state for display."""
        return {
            "word_display": self.word.get_display(),
            "wrong_guesses": self.wrong_guesses,
            "max_guesses": self.MAX_WRONG_GUESSES,
            "guessed_letters": sorted(self.word.guessed_letters),
            "hangman_stage": HangmanDisplay.get_stage(self.wrong_guesses),
            "game_over": self.game_over,
            "won": self.won,
        }


class GameInterface:
    """Handles user interface and interaction."""

    @staticmethod
    def clear_screen():
        """Clear the terminal screen."""
        os.system("cls" if os.name == "nt" else "clear")

    @staticmethod
    def display_header():
        """Display the game header."""
        print("\n" + "=" * 50)
        print("          HANGMAN GAME".center(50))
        print("=" * 50 + "\n")

    @staticmethod
    def display_game_state(game_state):
        """Display the current game state."""
        print(game_state["hangman_stage"])
        print(f"\nWord: {game_state['word_display']}")
        print(
            f"\nWrong guesses: {game_state['wrong_guesses']}/{game_state['max_guesses']}"
        )

        if game_state["guessed_letters"]:
            print(f"Guessed letters: {', '.join(game_state['guessed_letters'])}")
        else:
            print("Guessed letters: None yet")

    @staticmethod
    def display_player_stats(player):
        """Display player statistics."""
        stats = player.get_stats()
        print("\n" + "-" * 50)
        print(f"Player: {stats['name']}")
        print(
            f"Score: {stats['score']} | Games: {stats['games_played']} | Wins: {stats['games_won']}"
        )
        if stats["games_played"] > 0:
            print(f"Win Rate: {stats['win_rate']:.1f}%")
        print("-" * 50 + "\n")

    @staticmethod
    def get_user_input(prompt):
        """Get input from the user."""
        return input(prompt).strip()

    @staticmethod
    def display_message(message):
        """Display a message to the user."""
        print(f"\n>>> {message}")

    @staticmethod
    def display_welcome():
        """Display welcome message and rules."""
        GameInterface.clear_screen()
        print("\n" + "=" * 50)
        print("       WELCOME TO HANGMAN!".center(50))
        print("=" * 50)
        print("\nRULES:")
        print("  • Guess letters to reveal the hidden word")
        print("  • You have 6 wrong guesses before game over")
        print("  • Scoring: (6 - wrong_guesses) × 10 points")
        print("  • Enter 'quit' anytime to exit")
        print("\n" + "=" * 50)
        input("\nPress ENTER to start...")


def main():
    """Main game loop."""
    interface = GameInterface()
    interface.display_welcome()

    # Get player name
    interface.clear_screen()
    player_name = interface.get_user_input(
        "Enter your name (or press ENTER for 'Player'): "
    )
    player = Player(player_name if player_name else "Player")

    game = HangmanGame(player)
    playing = True

    while playing:
        # Start a new game
        game.start_new_game()

        while not game.game_over:
            interface.clear_screen()
            interface.display_header()
            interface.display_player_stats(player)

            game_state = game.get_game_state()
            interface.display_game_state(game_state)

            # Get user input
            guess = interface.get_user_input(
                "\nEnter a letter (or 'quit' to exit): "
            ).lower()

            if guess == "quit":
                playing = False
                break

            # Process the guess
            is_correct, message = game.make_guess(guess)

            if message:
                interface.clear_screen()
                interface.display_header()
                interface.display_player_stats(player)
                game_state = game.get_game_state()
                interface.display_game_state(game_state)
                interface.display_message(message)

                if game.game_over:
                    print()
                    input("Press ENTER to continue...")
                else:
                    import time

                    time.sleep(1.5)

        if not playing:
            break

        # Ask to play again
        interface.clear_screen()
        interface.display_player_stats(player)
        play_again = interface.get_user_input("Play again? (yes/no): ").lower()

        if play_again not in ["yes", "y"]:
            playing = False

    # Display final stats
    interface.clear_screen()
    print("\n" + "=" * 50)
    print("       THANKS FOR PLAYING!".center(50))
    print("=" * 50)
    interface.display_player_stats(player)
    print("\nGoodbye!\n")


if __name__ == "__main__":
    main()
